File: market_data/scripts/fix_xlsx.js
const XLSX = require('xlsx');
const fs = require('fs');
const path = require('path');

// Define the directory path
const directoryPath = '/Users/philipmassey/Downloads/';

// Read all files in the directory
fs.readdir(directoryPath, (err, files) => {
    if (err) {
        return console.error('Unable to scan directory: ' + err);
    }

    // Iterate over each file in the directory
    files.forEach(file => {
        // Check if the file has an .xlsx extension
        if (path.extname(file) === '.xlsx') {
            const filePath = path.join(directoryPath, file);
            //console.log(`Processing file: ${filePath}`);

            // Read the Excel file
            const workbook = XLSX.readFile(filePath);

            // Save the workbook, overwriting the original file
            XLSX.writeFile(workbook, filePath);

            console.log(`File has been overwritten: ${filePath}`);
        }
    });
});

File: market_data/scripts/fix_xlsx.py
import os
import pandas as pd

# Define the directory path
directory_path = os.path.expanduser('~/Downloads/')

# Iterate over each file in the directory
for file in os.listdir(directory_path):
    # Check if the file has an .xlsx extension
    if file.endswith('.xlsx'):
        file_path = os.path.join(directory_path, file)
        
        try:
            # Read all sheets using the 'calamine' engine which can handle invalid XML gracefully
            dfs = pd.read_excel(file_path, engine='calamine', sheet_name=None)
            
            # Save the workbook back, overwriting the original file using openpyxl (standard valid XML)
            with pd.ExcelWriter(file_path, engine='openpyxl') as writer:
                for sheet_name, df in dfs.items():
                    df.to_excel(writer, sheet_name=sheet_name, index=False)
            
            print(f"File has been successfully fixed and overwritten: {file_path}")
        except Exception as e:
            print(f"Error processing file {file_path}: {e}")
